### ISSUE TITLE
PUSH event hooks not sent for branch removal

### ISSUE THREAD

**Issue by stephenc** [go-gitea/gitea#2105]
- Gitea version (or commit ref): 858197b
- Git version: N/A
- Operating system: N/A
- Database (use `[x]`): N/A
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist:

## Description

When I configure a webhook that receives all events, push events that remove a branch or branch removal of the branch through the UI do not trigger a webhook event:

Creation of a new branch results in events like:

*X-Gitea-Event*: create
```
{
  "secret": "magic",
  "sha": "2409990415d77f47ed5c4c775abda4ffa6aa5f11",
  "ref": "branch",
  "ref_type": "branch",
  "repository": {
    "id": 17,
    "owner": {
...
```

and

*X-Gitea-Event*: push
```
{
  "secret": "magic",
  "ref": "refs/heads/branch",
  "before": "0000000000000000000000000000000000000000",
  "after": "2409990415d77f47ed5c4c775abda4ffa6aa5f11",
...
```

I would expect removal of the branch to trigger at least a webhook that looked something like:

*X-Gitea-Event*: push
```
{
  "secret": "magic",
  "ref": "refs/heads/branch",
  "before": "2409990415d77f47ed5c4c775abda4ffa6aa5f11",
  "after": "0000000000000000000000000000000000000000",
...
```

It is a somewhat separate concern the use of a magic value to indicate no ref, better would be to include the [created and deleted](https://developer.github.com/v3/activity/events/types/#pushevent) booleans that GitHub does, but for my part the only concern I have is the lack of a `push` event to catch branch removal.

**Comment by bkcsoft** [comment]
Could you include the logs from gitea? without them it's very hard to see what's happening :)

**Comment by bkcsoft** [comment]
(Not sure if this is bug or not, might be a feature request 😅 )

**Comment by stephenc** [comment]
I'll see if I can dig it out but your Jenkins integration (which I have nearly finished) will be annoying without an event for this.

If you want feature requests:

* manage hooks for a user's repos
* events for repo created / deleted
* some way in the api to distinguish a name as an org name vs user name (currently I get the name as a user and it's and org if the email is empty)
* file browse by rest api (basically the raw needs a way to serve directory listings for non-file paths... also good to flag content as a link if somebody commits a link)

Icing on top:

* commit file (with branch creation if desired) by rest api - needed for blue ocean integration

**Comment by stephenc** [comment]
Logs

Creating the branch: (`git push origin branch`)
```
Jul  3 20:05:56 sshd[27634]: Accepted publickey for git from x.x.x.x port 58096 ssh2: RSA SHA256:BTv+kWSXkHaRXdqhCNGet61YTXyMvWnfOFfSWr3Mgr8
[Macaron] 2017-07-03 20:05:56: Started GET /api/internal/branch/17/branch for 127.0.0.1
[Macaron] 2017-07-03 20:05:56: Completed /api/internal/branch/17/branch 200 OK in 2.354313ms
[Macaron] 2017-07-03 20:05:56: Started POST /api/internal/push/update for 127.0.0.1
[Macaron] 2017-07-03 20:05:56: Completed /api/internal/push/update 202 Accepted in 61.8573ms
[Macaron] 2017-07-03 20:05:56: Started POST /api/internal/ssh/1/update for 127.0.0.1
[Macaron] 2017-07-03 20:05:56: Completed /api/internal/ssh/1/update 200 OK in 13.464549ms
Jul  3 20:05:56 sshd[27634]: Received disconnect from x.x.x.x port 58096:11: disconnected by user
Jul  3 20:05:56 sshd[27634]: Disconnected from x.x.x.x port 58096
```

Events from the UI
<img width="464" alt="screen shot 2017-07-03 at 21 07 07" src="https://user-images.githubusercontent.com/209336/27806677-dc0471a4-6033-11e7-92c1-49993a4b062e.png">
<img width="446" alt="screen shot 2017-07-03 at 21 07 19" src="https://user-images.githubusercontent.com/209336/27806680-df35134c-6033-11e7-915b-5bb9380e2cdd.png">

Now the log when deleting the branch (`git push origin :branch`)

```
Jul  3 20:05:59 sshd[27690]: Accepted publickey for git from x.x.x.x port 58098 ssh2: RSA SHA256:BTv+kWSXkHaRXdqhCNGet61YTXyMvWnfOFfSWr3Mgr8
[Macaron] 2017-07-03 20:05:59: Started GET /api/internal/branch/17/branch for 127.0.0.1
[Macaron] 2017-07-03 20:05:59: Completed /api/internal/branch/17/branch 200 OK in 2.063422ms
[Macaron] 2017-07-03 20:05:59: Started POST /api/internal/push/update for 127.0.0.1
[Macaron] 2017-07-03 20:05:59: Completed /api/internal/push/update 202 Accepted in 9.581596ms
[Macaron] 2017-07-03 20:05:59: Started POST /api/internal/ssh/1/update for 127.0.0.1
[Macaron] 2017-07-03 20:05:59: Completed /api/internal/ssh/1/update 200 OK in 4.247786ms
Jul  3 20:05:59 sshd[27690]: Received disconnect from x.x.x.x port 58098:11: disconnected by user
Jul  3 20:05:59 sshd[27690]: Disconnected from x.x.x.x port 58098
```

No events in the UI, no events in request bin

@bkcsoft does that help?

**Comment by lunny** [comment]
I think that's a feature, see the description on the webhook on Gitea. There is no delete-branch.
![-2](https://user-images.githubusercontent.com/81045/27811805-82b56b9e-609c-11e7-9957-dd599e7a4643.png)

**Comment by stephenc** [comment]
Says "git push to a repository" so should be all git pushes, even those that remove refs.

That's a bug if you ask me.

A feature would be a separate new event for "delete" but the "push" should include removal based on its description

**Comment by stephenc** [comment]
@lunny 
<img width="187" alt="screen shot 2017-07-04 at 14 52 14" src="https://user-images.githubusercontent.com/209336/27832905-6547ba2c-60c8-11e7-92a9-f7db4b93d6d9.png">

Just says `Git push to a repository` it doesn't say `Subset of git pushes to a repository that add or update things but not those that remove things`

This is clearly a bug

**Comment by stephenc** [comment]
If you want to have a new event for non-push that is a delete event, that would be a separate feature request, but this issue is `PUSH event hooks not sent for branch removal`

**Comment by bkcsoft** [comment]
I would separate these checkboxes into categories, so you get somethings like this. WDYT?

- Git
  - Push event (new commits)
  - Branch events (create, delete)
  - Tag events (create, delete)
- PullRequest
  - Opened
  - Closed
  - [...]
- Issues
  - Opened
  - [...]

**Comment by stephenc** [comment]
I think push should align with the GitHub push event, i.e. Branch/tag CRUD

The other stuff would be a great feature request, but missing the D on this is a bug

**Comment by stephenc** [comment]
I also think that too many options is bad for users. GitHub has it about right:

<img width="677" alt="screen shot 2017-07-04 at 22 13 51" src="https://user-images.githubusercontent.com/209336/27842908-2ae1f0a8-6106-11e7-911b-d0d8b0c2ffa2.png">

You just need to fix this *bug* on how you process Push events... which should include pushes like `git push origin :branch` just like GitHub does.

The *PUSH* events currently are missing, clear bug.

The other events that GitHub offers are RFEs and should probably be in a separate issue

**Comment by stephenc** [comment]
@lunny @bkcsoft I have created https://github.com/go-gitea/gitea/issues/2113 to track the RFEs so that this one can be left to track the *BUG* ;-)

**Comment by bkcsoft** [comment]
If a hook is missing that never was there in the first place it's still a FE, not a bug ;)

**Comment by stephenc** [comment]
Well you already have a "Push" hook that promises "Git push to a repository" and I am showing that it does not deliver *all* git pushes to a repository... only the Cr & U of CrUD... so sounds very like a bug.... I am not asking for a new hook in this issue... only for the existing hook to publish all the "git push to a repository" that it currently advertises

Now you could fix this in one of two ways:

1. Actually fix the existing hook to honour its existing documentation
2. Change the documentation to say that "push" does not include the D in CrUD

Either of those are perfectly valid fixes to the bug... obviously I would be a Sad Panda if you chose the "change the documentation" route.

BTW when you fix this then https://github.com/jenkinsci/gitea-plugin will be a really sweet integration for gitea with Jenkins... right now, without the Push event publishing the D of CrUD you have to force scanning far far too often in order to clear out deleted branches

**Comment by daviian** [comment]
Someone's working on that?
Otherwise I would like to implement that.

**Comment by lafriks** [comment]
@daviian I don't think so, go ahead

**Comment by daviian** [comment]
IMO issue can be closed now as #2530 has been merged
